### ISSUE TITLE
[css-lists-3] list-item counter nesting is confusing

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#9076]
See @ZoeBijl’s thread at https://w3c.social/@moiety@front-end.social/110729923581670564

The testcase uses the following code:

```
<ol>
    <li>Some</li>
    <li>List</li>
    <li>Items
        <ol>
        <li>Nested</li>
        </ol>
    </li>
</ol>

<section>
    <ol>
        <li>Some</li>
        <li>List</li>
        <li>Items
            <ol>
                <li>Nested</li>
            </ol>
        </li>
    </ol>
</section>

<ol>
    <li>Some</li>
    <li>List
        <ol>
            <li>Nested</li>
        </ol>
    </li>
    <li>Items
    </li>
</ol>
```

and using the `counters()` function results in what looks like nonsense (that is apparently nonetheless per spec). I'm not sure how to fix it, but we should probably fix it... it might require some new feature that adjusts counter scoping.

**Comment by dbaron** [comment]
Maybe we should introduce the concept of "narrow" versus "wide" scope to counters, where narrow means element + descendants, and wide means element + later siblings + descendants of all.  Counters currently use wide scope, but lists would be better served by narrow scope.  (This concept is explicit in the [CSS Toggles proposal](https://tabatkins.github.io/css-toggle/) and probably in other things.)

Counters were designed to try to handle both lists and headers (h1...h6)... and the original spec had some errors.  When I initially implemented in in Gecko, I proposed fixes for the errors in a way that would make the key use cases work, but in hindsight it would have been better to propose bigger structural changes rather than the minimal thing to turn the then-current spec draft into something that actually rendered its examples as promised.

**Comment by Loirooriol** [comment]
I think narrow scoped counters would also allow fixing https://github.com/w3c/csswg-drafts/issues/5477 in a proper way, and I support that.

**Comment by Loirooriol** [comment]
Just want to note that there are old sites relying on the "nonsense" (expecting that `counter-reset` will propagate to following siblings). That's because we didn't have `counter-set` in the past, so people made do with `counter-reset`. So changing the current behavior doesn't seem possible due to compat, would need to be something new like `counter-reset: scoped(counter-name) 5`

**Comment by tabatkins** [comment]
Yeah this is definitely not changeable by default, counters must have wide scope. But making it narrow scope optionally is a good idea!

**Comment by tabatkins** [comment]
Note, the rendering of the OP's example is:

```
1. Some
2. List
3. Items
    3.1. Nested
3.1 Some
3.2 List
3.3 Items
    3.3.1. Nested
1. Some
2. List
    2.1. Nested
3. Items
```

**Comment by mirisuzanne** [comment]
Might it be possible to change the default only for the special `list-item` counter, which is a much more recent addition? I don't believe this is how list item counters have worked historically.

**Comment by tabatkins** [comment]
I think historically it wasn't possible to tell what the behavior was, since you couldn't refer to the `list-item` counter for `counters()`. (The behavior is unobservable if you're just looking at the value of the innermost counter for a given name.)

But yeah it is likely possible to change the HTML default.

**Comment by Loirooriol** [comment]
Yeah, the UA stylesheet has something like `ol, ul { counter-reset: list-item }`. It's hopefully fine to upgrade this to `ol, ul { counter-reset: sane-scope(list-item) }`, what can't change is the behavior of `counter-reset: list-item`.

**Comment by Loirooriol** [comment]
Just noting that this isn't only for `counters()`, basic HTML lists also have oddities as currently defined (since  #5477 was not a comprehensive solution):

<details>

```html
<ol>
    <li>Some</li>
    <li>List</li>
    <li>Items
        <ol>
            <li>Nested</li>
        </ol>
    </li>
</ol>
<section style="list-style-type: decimal">
    <li>Some</li>
    <li>List</li>
    <li>Items
        <ol>
            <li>Nested</li>
        </ol>
    </li>
</section>
<ol>
    <li>Some</li>
    <li>List
        <ol>
            <li>Nested</li>
        </ol>
    </li>
    <li>Items</li>
</ol>
```

Blink and WebKit don't use actual counters, so they show the more reasonable

```
1. Some
2. List
3. Items
    1. Nested

1 Some
2 List
3 Items
    1. Nested

1. Some
2. List
    1. Nested
3. Items
```

But in Firefox, or when manually switching to an actual counter,


```
1. Some
2. List
3. Items
    1. Nested

4 Some
5 List
6 Items
    1. Nested

1. Some
2. List
    1. Nested
3. Items
```

</details>

**Comment by Loirooriol** [comment]
I guess we will need 2 variants of narrow-scoped counters: normal and reversed. `scoped()` and `scoped-reversed()`?

**Comment by tabatkins** [comment]
Something like that, yes.

**Comment by Loirooriol** [comment]
I guess the basic idea would be that [counters](https://drafts.csswg.org/css-lists/#counter) would have a bool to track if they are narrow-scoped. Then, [counter inheritance](https://drafts.csswg.org/css-lists/#inherit-counters) would skip counters from the preceding sibling which are narrow-scoped and whose creator is the preceding sibling.

What I'm less sure about is whether we want to preserve this behavior of HTML ordinals in Blink:

```html
<div style="list-style: decimal; padding-left: 40px">
  <li></li>
  <li></li>
  <ol>
    <li></li>
    <li></li>
    <li></li>
  </ol>
  <li></li>
  <li></li>
</div>
```
```
1.
2.
    1.
    2.
    3.
3.
4.
```

If so, when skipping a counter as described above, would need to iterate the previous siblings until finding one which either doesn't have a counter with that name, or doesn't have the narrow-scoped flag, or the creator is an ancestor.

But it may be fine to produce
```
1.
2.
    1.
    2.
    3.
1.
2.
```
and just require instantiating the counter on the parent for the earlier behavior (like using `<ol>` instead of `<div>`).

**Comment by astearns** [comment]
See discussion in https://github.com/w3c/csswg-drafts/issues/5477#issuecomment-1843956580

**Comment by Loirooriol** [comment]
Remaining details to decide:

 - Name: `scoped()`, `narrow-scoped()`?
 - Combination with `reversed()`: another function like `scoped-reversed()` or nesting `scoped(reversed())`
 - If there is a non-scoped counter with the same name whose creator is an earlier sibling,
   1. Whether `scoped()` instantiates a new counter in addition to or in place of the earlier one.
       Maybe "in place of", like when instantiating non-scoped counters, but this can impact later siblings...
   2. If "in place of", whether later siblings can still inherit the earlier counter or not.
       Not inheriting seems simpler, but inheriting is maybe a bit better and matches HTML ordinals in Blink and WebKit.

**Comment by fantasai** [comment]
I think instead of making new function types, it might be better to introduce an additional property e.g. `counter-scope: [ narrow | wide ]#` or something like that?

**Comment by Loirooriol** [comment]
HTML ordinals should use narrow scopes. If there is a single property controlling the scope of all counters, it can break websites if they are using their own counters on lists.

**Comment by Loirooriol** [comment]
Well, I guess we could avoid the risk by saying that a counter is scoped if its creator has `counter-scope: narrow` or if its name is `list-item`. This would make `list-item` more magical, but it would be fine.

**Comment by fantasai** [comment]
Alternate keywords: `counter-scope: contain | cover` where `contain` scopes the counter declared with `counter-reset` to the contents of the element (not the element itself).
